File: python/monkey/parser.py
from typing import Callable
from enum import Enum, auto
import monkey.ast as ast
import monkey.token_types as TokenType
from monkey.lexer import Lexer
from monkey.token import Token

prefix_parse_func = Callable[[], ast.Expression]
infix_parse_func = Callable[[ast.Expression], ast.Expression]

class Precedence(Enum):
    LOWEST = auto()
    EQUALS = auto()  # ==
    LESSGREATER = auto()  # > or <
    SUM = auto()  # +
    PRODUCT = auto()  # *
    PREFIX = auto()  # -X or !X
    CALL = auto()  # myFunction(X)

class Parser:
    def __init__(self, lexer: Lexer) -> None:
        self.lexer: Lexer = lexer
        self.current_token: Token = None
        self.peek_token: Token = None
        self.errors: list[str] = []

        self.prefix_parse_funcs : dict[str, prefix_parse_func] = {}
        self.infix_parse_funcs : dict[str, infix_parse_func] = {}

        self.register_prefix(TokenType.IDENT, self.parse_identifier)
        self.register_prefix(TokenType.INT, self.parse_integer_literal)
        self.register_prefix(TokenType.BANG, self.parse_prefix_expression)
        self.register_prefix(TokenType.MINUS, self.parse_prefix_expression)

        self.next_token()
        self.next_token()

    def next_token(self) -> None:
        self.current_token = self.peek_token
        self.peek_token = self.lexer.get_next_token()

    def register_prefix(self, token_type, func : prefix_parse_func) -> None:
        self.prefix_parse_funcs[token_type] = func
    
    def register_infix(self, token_type, func : infix_parse_func) -> None:
        self.infix_parse_funcs[token_type] = func
    
    def parse_program(self) -> ast.Program:
        program = ast.Program()

        while self.current_token.type != TokenType.EOF:
            stmt = self.parse_statement()
            if stmt:
                program.statements.append(stmt)
            self.next_token()
        
        return program

    def parse_identifier(self) -> ast.Expression:
        return ast.Identifier(token=self.current_token, value= self.current_token.literal)
    
    def parse_integer_literal(self) -> ast.Expression | None:
        int_lit = ast.IntegerLiteral(token=self.current_token)
        if not self.current_token.literal.isdigit():
            msg = f"Could not parse {self.current_token.literal} as digit"
            self.errors.append(msg)
            return None

        int_lit.value = int(self.current_token.literal)
        return int_lit
    
    def parse_prefix_expression(self) -> ast.Expression:
        expression = ast.PrefixExpression(token=self.current_token, operator= self.current_token.literal)

        self.next_token()
        expression.right = self.parse_expression(Precedence.PREFIX)
        return expression
    
    def parse_statement(self) -> ast.Statement:
        match self.current_token.type:
            case TokenType.LET:
                return self.parse_let_statement()
            case TokenType.RETURN:
                return self.parse_return_statement()
            case _:
                return self.parse_expression_statement()
            
    def parse_let_statement(self) -> ast.LetStatement:
        stmt = ast.LetStatement(token=self.current_token)

        if not self.expect_peek(TokenType.IDENT):
            return None
        
        stmt.name = ast.Identifier(token= self.current_token, value= self.current_token.literal)

        if not self.expect_peek(TokenType.ASSING):
            return None
        
        # Skip until semicolon
        while self.current_token.type != TokenType.SEMICOLON:
            self.next_token()

        return stmt
    
    def parse_return_statement(self) -> ast.ReturnStatement:
        stmt = ast.ReturnStatement(token=self.current_token)
        
        self.next_token()
        while self.current_token.type != TokenType.SEMICOLON:
            self.next_token()
        
        return stmt
    
    def parse_expression_statement(self) -> ast.ExpressionStatement:
        stmt = ast.ExpressionStatement(token=self.current_token)

        stmt.expression = self.parse_expression(Precedence.LOWEST)

        self.expect_peek(TokenType.SEMICOLON)

        return stmt
    
    def parse_expression(self, precedence: Precedence) -> ast.Expression:
        prefix = self.prefix_parse_funcs.get(self.current_token.type, None)

        if not prefix:
            return None
        
        left_exp = prefix()
        return left_exp

    
    def peek_token_is(self, type ) -> bool:
        return self.peek_token.type == type

    def expect_peek(self,type) -> bool:
        if self.peek_token_is(type):
            self.next_token()
            return True
        self.peek_error(type)
        return False
    
    def peek_error(self, type) -> None:
        msg = "Expected next token to be %s, got %s instead" % (type, self.peek_token.type)
        self.errors.append(msg)

File: java/monkey/src/main/java/root/parser/Parser.java
package root.parser;

import root.LocalizedToken;
import root.TokenType;
import root.ast.Program;
import root.ast.expressions.*;
import root.ast.statements.*;
import root.lexer.Lexer;

import java.util.ArrayList;
import java.util.List;

public class Parser {

    private final Lexer lexer;

    private LocalizedToken currentToken;

    private LocalizedToken peekToken;

    private final List<ParserException> errors = new ArrayList<>();

    public Parser(Lexer lexer) {
        this.lexer = lexer;

        proceedToNextToken();
        proceedToNextToken();
    }

    public Program parseProgram() throws ParseProgramException {
        var program = new Program();

        while (!currentTokenIs(TokenType.EOF)) {
            try {
                Statement statement = parseStatement();
                if (statement != null) {
                    program.getStatements().add(statement);
                }
            } catch (ParserException pe) {
                errors.add(pe);
            }
            proceedToNextToken();
        }

        if (!errors.isEmpty()) {
            throw new ParseProgramException(errors);
        }

        return program;
    }

    private void proceedToNextToken() {
        currentToken = peekToken;
        peekToken = lexer.nextLocalized();
    }

    private Statement parseStatement() throws ParserException {
        return switch (currentToken.type()) {
            case LET -> parseLetStatement();
            case RETURN -> parseReturnStatement();
            default -> parseExpressionStatement();
        };
    }

    private ReturnStatement parseReturnStatement() throws ParserException {
        var returnStatement = new ReturnStatement(currentToken);

        proceedToNextToken();

        // A modification I made to support the new Unit type for functions that don't return anything
        if (currentTokenIs(TokenType.SEMI)) {
            returnStatement.setReturnValue(UnitExpression.INSTANCE);
            return returnStatement;
        }

        returnStatement.setReturnValue(parseExpression(OperatorPrecedence.LOWEST));

        if (peekTokenIs(TokenType.SEMI)) {
            proceedToNextToken();
        }

        return returnStatement;
    }

    private LetStatement parseLetStatement() throws ParserException {
        var letStatement = new LetStatement(currentToken);

        expectPeek(TokenType.IDENTIFIER);

        letStatement.setName(new IdentifierExpression(currentToken, currentToken.literal()));

        expectPeek(TokenType.ASSIGN);
        proceedToNextToken();

        letStatement.setValue(parseExpression(OperatorPrecedence.LOWEST));

        if (peekTokenIs(TokenType.SEMI)) {
            proceedToNextToken();
        }

        return letStatement;
    }

    private ExpressionStatement parseExpressionStatement() throws ParserException {
        var statement = new ExpressionStatement(currentToken);

        statement.setExpression(parseExpression(OperatorPrecedence.LOWEST));

        if (peekTokenIs(TokenType.SEMI)) {
            proceedToNextToken();
        }

        return statement;
    }

    private Expression parseExpression(OperatorPrecedence precedence) throws ParserException {
        ParserSupplier<Expression> prefixFn = prefixParseFn();

        if (prefixFn == null) {
            throw new ParserException("Unexpected token found: " + currentToken.literal(), currentToken);
        }
        Expression leftExpression = prefixFn.get();

        while (!peekTokenIs(TokenType.SEMI) && precedence.compareTo(peekPrecedence()) < 0) {
            ParserFunction<Expression, Expression> infixFn = infixParseFn(peekToken.type());

            if (infixFn == null) {
                break;
            }

            proceedToNextToken();

            leftExpression = infixFn.apply(leftExpression);
        }

        return leftExpression;
    }

    private ParserSupplier<Expression> prefixParseFn() {
        return switch (currentToken.type()) {
            case IDENTIFIER -> () -> new IdentifierExpression(currentToken, currentToken.literal());
            case INT -> () -> new IntegerLiteralExpression(currentToken, Long.parseLong(currentToken.literal()));
            case TRUE, FALSE -> () -> new BooleanLiteralExpression(currentToken, currentTokenIs(TokenType.TRUE));
            case NULL -> () -> new NullLiteralExpression(currentToken);
            case STRING -> () -> new StringLiteralExpression(currentToken);
            case BANG, MINUS -> this::parsePrefixExpression;
            case LPAREN -> this::parseGroupedExpression;
            case IF -> this::parseIfExpression;
            case FUNC -> this::parseFunctionLiteral;
            case LBRACKET -> this::parseArrayLiteralExpression;
            default -> null;
        };
    }

    private PrefixExpression parsePrefixExpression() throws ParserException {
        var expression = new PrefixExpression(currentToken, currentToken.literal());
        proceedToNextToken();
        expression.setRight(parseExpression(OperatorPrecedence.PREFIX));
        return expression;
    }

    private ParserFunction<Expression, Expression> infixParseFn(TokenType tokenType) {
        return switch (tokenType) {
            case PLUS, MINUS, SLASH, ASTERISK, EQUAL, NOT_EQUAL, LT, GT, AND, OR -> this::parseInfixExpression;
            case LPAREN -> this::parseCallExpression;
            case LBRACKET -> this::parseIndexExpression;
            default -> null;
        };
    }

    private InfixExpression parseInfixExpression(Expression left) throws ParserException {
        var expression = new InfixExpression(currentToken, currentToken.literal(), left);

        var precedence = curPrecedence();
        proceedToNextToken();
        expression.setRight(parseExpression(precedence));

        return expression;
    }

    private CallExpression parseCallExpression(Expression function) throws ParserException {
        LocalizedToken callToken = currentToken;
        List<Expression> arguments = parseExpressionList(TokenType.RPAREN);

        return new CallExpression(callToken, function, arguments);
    }

    private Expression parseIndexExpression(Expression left) throws ParserException {
        LocalizedToken openBracketToken = currentToken;

        proceedToNextToken();

        Expression index = parseExpression(OperatorPrecedence.LOWEST);

        expectPeek(TokenType.RBRACKET);

        return new IndexExpression(openBracketToken, left, index);
    }

    private List<Expression> parseExpressionList(TokenType limitingToken) throws ParserException {
        var expressions = new ArrayList<Expression>();

        if (peekTokenIs(limitingToken)) {
            proceedToNextToken();
            return expressions;
        }

        while (true) {
            proceedToNextToken();
            var argument = parseExpression(OperatorPrecedence.LOWEST);
            expressions.add(argument);
            if (peekTokenIs(TokenType.COMMA)) {
                proceedToNextToken();
            } else {
                break;
            }
        }

        expectPeek(limitingToken);

        return expressions;
    }

    private Expression parseGroupedExpression() throws ParserException {
        proceedToNextToken();

        Expression expression = parseExpression(OperatorPrecedence.LOWEST);

        expectPeek(TokenType.RPAREN);

        return expression;
    }

    private IfExpression parseIfExpression() throws ParserException {
        var ifExpression = new IfExpression(currentToken);

        expectPeek(TokenType.LPAREN);
        proceedToNextToken();

        ifExpression.setCondition(parseExpression(OperatorPrecedence.LOWEST));

        expectPeek(TokenType.RPAREN);
        expectPeek(TokenType.LSQIRLY);

        ifExpression.setConsequence(parseBlockStatement());

        if (peekTokenIs(TokenType.ELSE)) {
            proceedToNextToken();
            expectPeek(TokenType.LSQIRLY);

            ifExpression.setAlternative(parseBlockStatement());
        }

        return ifExpression;
    }

    private FunctionLiteralExpression parseFunctionLiteral() throws ParserException {
        var functionLiteral = new FunctionLiteralExpression(currentToken);

        expectPeek(TokenType.LPAREN);

        parseFunctionParameters(functionLiteral);

        expectPeek(TokenType.LSQIRLY);

        functionLiteral.setBody(parseBlockStatement());

        return functionLiteral;
    }

    private void parseFunctionParameters(FunctionLiteralExpression functionLiteral) throws ParserException {
        if (peekTokenIs(TokenType.RPAREN)) {
            proceedToNextToken();
            return;
        }

        expectPeek(TokenType.IDENTIFIER);
        var identifier = new IdentifierExpression(currentToken, currentToken.literal());
        functionLiteral.getParameters().add(identifier);

        if (peekTokenIs(TokenType.COMMA)) {
            proceedToNextToken();
            parseFunctionParameters(functionLiteral);
        } else {
            expectPeek(TokenType.RPAREN);
        }
    }

    private Expression parseArrayLiteralExpression() throws ParserException {
        LocalizedToken openBracketToken = currentToken;
        List<Expression> elements = parseExpressionList(TokenType.RBRACKET);

        return new ArrayLiteralExpression(openBracketToken, elements);
    }

    private BlockStatement parseBlockStatement() throws ParserException {
        var blockStatement = new BlockStatement(currentToken);

        proceedToNextToken();

        while (!currentTokenIs(TokenType.RSQIRLY) && !currentTokenIs(TokenType.EOF)) {
            blockStatement.addStatement(parseStatement());
            proceedToNextToken();
        }

        if (!currentTokenIs(TokenType.RSQIRLY)) {
            throw new ParserException("Block statement needs to be closed", currentToken);
        }

        return blockStatement;
    }

    private boolean currentTokenIs(TokenType type) {
        return currentToken.type() == type;
    }

    private OperatorPrecedence curPrecedence() {
        return OperatorPrecedence.precedenceForTokenType(currentToken.type());
    }

    private boolean peekTokenIs(TokenType type) {
        return peekToken.type() == type;
    }

    private OperatorPrecedence peekPrecedence() {
        return OperatorPrecedence.precedenceForTokenType(peekToken.type());
    }

    private void expectPeek(TokenType type) throws ParserException {
        if (peekTokenIs(type)) {
            proceedToNextToken();
        } else {
            throw new ParserException("Expected next token to be %s, got %s".formatted(type.tokenOrName(), peekToken.literal()), currentToken);
        }
    }
}
